### PR TITLE
Support out argument in torch.fft ops

### PR DESCRIPTION
**Pull request by peterbell10**
Ref #42175

This adds out argument support to all functions in the `torch.fft` namespace except for `fftshift` and `ifftshift` because they rely on `at::roll` which doesn't have an out argument version. 

Note that there's no general way to do the transforms directly into the output since both cufft and mkl-fft only support single batch dimensions. At a minimum, the output may need to be re-strided which I don't think is expected from `out` arguments normally. So, on cpu this just copies the result into the out tensor. On cuda, the normalization is changed to call `at::mul_out` instead of an inplace multiply.

If it's desirable, I could add a special case to transform into the output when `out.numel() == 0` since there's no expectation to preserve the strides in that case anyway. But that would lead to the slightly odd situation where `out` having the correct shape follows a different code path from `out.resize_(0)`.

### CODE REVIEW

File: aten/src/ATen/native/SpectralOps.cpp [hunk 1]
```diff
@@ -102,9 +102,12 @@ Tensor resize_fft_input(Tensor x, IntArrayRef dims, IntArrayRef sizes) {
 }
 
 // Complex to real FFT
-Tensor fft_c2r(Tensor input, c10::optional<int64_t> n_opt,
+Tensor fft_c2r(Tensor out, Tensor input, c10::optional<int64_t> n_opt,
                int64_t unwrapped_dim, c10::optional<std::string> norm_str,
                bool forward) {
+  TORCH_CHECK(!out.defined() || out.is_floating_point(), forward ? "ih" : "ir",
+              "fft expects a (real) floating point output tensor, but got ",
```
Comment (mruberry) [inline aten/src/ATen/native/SpectralOps.cpp:109]: This can probably ditch the "(real)" since "floating point" is specified.

File: aten/src/ATen/native/SpectralOps.cpp [hunk 2]
```diff
@@ -136,16 +145,24 @@ Tensor fft_r2c(Tensor input, c10::optional<int64_t> n_opt,
...
+  if (out.defined() && forward) {
+    ret = at::_fft_r2c_out(out, input, dim, static_cast<int64_t>(norm), onesided);
+  } else {
+    ret = at::_fft_r2c(input, dim, static_cast<int64_t>(norm), onesided);
+  }
+
   if (!forward) {
     // FIXME: _fft_r2c doesn't support native r2c IFFT
-    out = at::conj(out);
+    return out.defined() ? at::conj_out(out, ret) : at::conj(ret);
```
Comment (mruberry) [inline aten/src/ATen/native/SpectralOps.cpp:158]: In the case that out is defined, ret == out, right? Would it be more natural to perform the conj inplace instead of using its out variant? 

This question is mostly academic since conj doesn't have an inplace variant. cc @anjali411 we should probably add an inplace `conj_` for consistency with our other operators?

File: aten/src/ATen/native/SpectralOps.cpp [hunk 3]
```diff
@@ -230,12 +252,17 @@ ShapeAndDims canonicalize_fft_shape_and_dim_args(
...
-    const Tensor& input, IntArrayRef shape, IntArrayRef dim,
-    c10::optional<std::string> norm_str, bool forward) {
+    Tensor out, const Tensor& input, IntArrayRef shape,
+    IntArrayRef dim, c10::optional<std::string> norm_str, bool forward) {
   TORCH_CHECK(input.is_complex(), "Expected a complex input tensor to FFT");
   Tensor x = resize_fft_input(input, dim, shape);
   const auto norm = norm_from_string(norm_str, forward);
-  return at::_fft_c2c(x, dim, static_cast<int64_t>(norm), forward);
+  if (out.defined()) {
+    TORCH_CHECK(out.is_complex(), "fft expects a complex output tensor, but got ", out.scalar_type());
```
Comment (mruberry) [inline aten/src/ATen/native/SpectralOps.cpp:261]: "fft expects" -> "fftn expects"?

File: aten/src/ATen/native/SpectralOps.cpp [hunk 4]
```diff
@@ -323,49 +436,97 @@ Tensor fft_irfftn(const Tensor& self, c10::optional<IntArrayRef> s,
   Tensor input = promote_tensor_fft(self, /*require_complex=*/true);
   Tensor x = resize_fft_input(input, desc.dim, desc.shape);
   const auto norm = norm_from_string(norm_str, /*forward=*/false);
-  return at::_fft_c2r(x, desc.dim, static_cast<int64_t>(norm), last_dim_size);
+  if (out.defined()) {
+    TORCH_CHECK(out.is_floating_point(), "irfftn expects a (real) floating point output tensor, but got ", out.scalar_type());
```
Comment (mruberry) [inline aten/src/ATen/native/SpectralOps.cpp:440]: Consistent with above comment, can probably remove "(real)" here

File: aten/src/ATen/native/SpectralOps.cpp [hunk 5]
```diff
@@ -157,7 +174,12 @@ Tensor fft_c2c(Tensor input, c10::optional<int64_t> n_opt,
     input = resize_fft_input(input, dim, n);
   }
   const auto norm = norm_from_string(norm_str, forward);
-  return at::_fft_c2c(input, dim, static_cast<int64_t>(norm), forward);
+  if (out.defined()) {
+    TORCH_CHECK(out.is_complex(), "fft expects a complex output tensor, but got ", out.scalar_type());
```
Comment (mruberry) [inline aten/src/ATen/native/SpectralOps.cpp:178]: "fft" here is interesting because this warning could be triggered by many operations. 

In the above warning a capital "FFT" is used, which doesn't seem much better.

Maybe this function should accept a string with the appropriate function name to display? So it could say, for example, "fft2 expects..." when `torch.fft.fft2` is called?

I guess c2r and r2c already have enough logic to disambiguate which function called them, but they could adopt this approach for consistency, too.

File: aten/src/ATen/native/SpectralOps.cpp (hunk 3 above)
Comment (mruberry) [reply aten/src/ATen/native/SpectralOps.cpp:261]: Actually, see comment above about passing a string to specialize the error message appropriately.

File: aten/src/ATen/native/SpectralOps.cpp [hunk 6]
```diff
@@ -323,49 +436,97 @@ Tensor fft_irfftn(const Tensor& self, c10::optional<IntArrayRef> s,
...
-  ScalarType dtype = typeMetaToScalarType(options.dtype());
+Tensor& fft_irfft2_out(Tensor& out, const Tensor& self, c10::optional<IntArrayRef> s,
+                       IntArrayRef dim, c10::optional<std::string> norm) {
+  return native::fft_irfftn_out(out, self, s, dim, std::move(norm));
+}
+
+Tensor& fft_fftfreq_out(Tensor& out, int64_t n, double d) {
+  ScalarType dtype = out.scalar_type();
   TORCH_CHECK(at::isFloatingType(dtype) || at::isComplexType(dtype),
               "fftfreq requires a floating point or complex dtype");
```
Comment (mruberry) [inline aten/src/ATen/native/SpectralOps.cpp:505]: What are the error messages for other tensor creation ops, like `torch.arange` and `torch.zeros`, like for the out vs tensor options case? What happens if both out= and some tensor options are specified?

File: aten/src/ATen/native/native_functions.yaml [hunk 7]
```diff
@@ -2062,6 +2062,12 @@
     CPU: _fft_r2c_mkl
     CUDA: _fft_r2c_cufft
 
+- func: _fft_r2c.out(Tensor self, int[] dim, int normalization, bool onesided, *, Tensor(a!) out) -> Tensor(a!)
```
Comment (mruberry) [inline aten/src/ATen/native/native_functions.yaml:2065]: I imagine this will run afoul of https://github.com/pytorch/pytorch/pull/49510 and need an update.

cc @smessmer

File: torch/csrc/utils/out_types.cpp [hunk 8]
```diff
@@ -0,0 +1,37 @@
+#include <torch/csrc/utils/out_types.h>
+
+namespace torch {
+namespace utils {
+
+void check_out_type_matches(const at::Tensor& result,
```
Comment (mruberry) [inline torch/csrc/utils/out_types.cpp:6]: Add a comment describing this function & how it's intended to be used

File: torch/csrc/utils/out_types.cpp [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+                            at::ScalarType scalarType, bool scalarType_is_none,
+                            c10::optional<at::Layout> layout,
+                            const at::Device& device, bool device_is_none) {
+  if (scalarType_is_none && !layout && device_is_none) {  // common case
+    return;
+  }
+  if (!scalarType_is_none && result.scalar_type() != scalarType) {
+    AT_ERROR(
```
Comment (mruberry) [inline torch/csrc/utils/out_types.cpp:14]: AT_ERROR is deprecated in favor of TORCH_CHECK (for errors that a user input can trigger) or TORCH_INTERNAL_ASSERT (for internal logic failures that should never occur).

File: torch/csrc/utils/out_types.cpp [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+        "dtype ", scalarType,
```
Comment (mruberry) [inline torch/csrc/utils/out_types.cpp:15]: This will probably also need to be passed a string to produce an error message associating the calling function with the error

File: torch/csrc/utils/out_types.cpp (hunk 8 above)
Comment (mruberry) [reply torch/csrc/utils/out_types.cpp:6]: What's the thinking for moving this to a new file?

File: test/test_spectral_ops.py [hunk 11]
```diff
@@ -623,6 +649,19 @@ def test_fftfreq_numpy(self, device, dtype):
                 actual = torch_fn(*args, device=device, dtype=dtype)
                 self.assertEqual(actual, expected, exact_dtype=False)
 
+    @skipCPUIfNoMkl
+    @skipCUDAIfRocm
+    @onlyOnCPUAndCUDA
+    @dtypes(torch.float, torch.double)
+    def test_fftfreq_out(self, device, dtype):
```
Comment (mruberry) [inline test/test_spectral_ops.py:656]: Seems like we should make these OpInfos in the (near) future?

File: test/test_spectral_ops.py [hunk 12]
```diff
@@ -332,6 +332,32 @@ def test_fft_backward(self, device, dtype):
...
+    def test_fft_invalid_out_types(self, device):
+
+        complex_fft_funcs = [torch.fft.fft, torch.fft.ifft, torch.fft.fftn, torch.fft.ifftn,
+                             torch.fft.rfft, torch.fft.rfftn, torch.fft.ihfft]
+        real_fft_funcs = [torch.fft.irfft, torch.fft.irfftn, torch.fft.hfft]
+        fft_funcs = complex_fft_funcs + real_fft_funcs
+
+        # Test errors on invalid out dtypes
+        x = torch.rand(10, device=device, dtype=torch.float32)
+        for out_dtype in (torch.int16, torch.float32, torch.complex64):
```
Comment (mruberry) [inline test/test_spectral_ops.py:347]: Code golf: this could just be an iterable of pairs like (out, dtype, funcs to test):

```
for out_dtype, funcs in ((torch.int16, fft_funcs), ...):
```

**Review by mruberry — COMMENTED** [review]
Hey @peterbell10! Overall this looks great, as usual. Just a few comments (inline) and (probably) a merge conflict to resolve. 

Let me know if you have questions about updating the dispatch metadata. @smessmer - we should probably update the ATen/native README?

File: test/test_spectral_ops.py (hunk 11 above)
Comment (peterbell10) [reply test/test_spectral_ops.py:656]: Are there any `OpInfos` for tensor creation functions yet?

File: aten/src/ATen/native/SpectralOps.cpp (hunk 6 above)
Comment (peterbell10) [reply aten/src/ATen/native/SpectralOps.cpp:505]: >What happens if both out= and some tensor options are specified?

The function I moved out of `python_torch_functions.cpp` is automatically called by codegen to ensure the tensor options arguments are valid. For some reason this happens even though I don't take any `TensorOptions` arguments here.

Comment (mruberry) [reply aten/src/ATen/native/SpectralOps.cpp:505]: That's odd. Don't other functions, like those mentioned above, have this same issue? What's the implication of this function always being called even though no TensorOptions are in the signature?

Comment (mruberry) [reply aten/src/ATen/native/SpectralOps.cpp:505]: Or are the `fftfreq` functions the first tensor factories to hit some unique aspect of this issue?

File: test/test_spectral_ops.py (hunk 11 above)
Comment (mruberry) [reply test/test_spectral_ops.py:656]: Have to start somewhere.

File: aten/src/ATen/native/SpectralOps.cpp (hunk 6 above)
Comment (peterbell10) [reply aten/src/ATen/native/SpectralOps.cpp:505]: >What's the implication of this function always being called even though no TensorOptions are in the signature?

The non-out variant uses `TensorOptions` to create the out tensor; the out variant respects the `TensorOptions` of the `out` tensor.  And the python API ensures that if both are given, they must be consistent.

>Or are the fftfreq functions the first tensor factories to hit some unique aspect of this issue?

I don't think it's unique, I think it's just a quirk of the python codegen. For example, I see that `empty.out` doesn't have a `dtype` argument yet I can still do `torch.empty(100, out=out, dtype=torch.float)`.

There is also no `TensorOptions` argument in the native function:
https://github.com/pytorch/pytorch/blob/4883d39c6fd38431bdc60e1db6402e251429b1e1/aten/src/ATen/native/TensorFactories.cpp#L198-L201

Comment (mruberry) [reply aten/src/ATen/native/SpectralOps.cpp:505]: That's really interesting. cc @bhosmer for yet another quirk of our current codegen.

File: aten/src/ATen/native/SpectralOps.cpp (hunk 5 above)
Comment (peterbell10) [reply aten/src/ATen/native/SpectralOps.cpp]: I've added the name argument but  `fft2` will still say `fftn` in error messages because I'm calling `native::fftn` rather than `fftn_c2c` which expects slightly different arguments.

Comment (mruberry) [reply aten/src/ATen/native/SpectralOps.cpp]: I think we can live with that.

File: aten/src/ATen/native/native_functions.yaml (hunk 7 above)
Comment (peterbell10) [reply aten/src/ATen/native/native_functions.yaml:2176]: When I wrote this I ha to drop `use_c10_dispatcher: full` because the out argument can't go at the end since it doesn't have a default. Looks like `hacky_wrapper_for_legacy_signatures` works fine though and it compiles after rebasing on #49510.

File: aten/src/ATen/native/SpectralOps.cpp (hunk 2 above)
Comment (anjali411) [reply aten/src/ATen/native/SpectralOps.cpp:160]: @mruberry yeah I agree I don't see any reason to not add an in-place version of conj

File: torch/csrc/utils/out_types.cpp (hunk 10 above)
Comment (mruberry) [reply torch/csrc/utils/out_types.cpp:17]: OK.

**Comment by mruberry** [comment]
Hey @peterbell10, just checking in on this. Is this gtg?

**Comment by peterbell10** [comment]
Yes @mruberry, this should be gtg.

**Review by mruberry — APPROVED** [review]
Cool! Thanks @peterbell10!
